Lib/glyphsLib/builder/transformations/intermediate_layers.py

```diff
@@ -0,0 +1,118 @@
+import logging
+import uuid
+
+from fontTools.varLib.models import VariationModel, normalizeValue
+
+from glyphsLib.classes import GSLayer, GSNode, GSPath
+from glyphsLib.builder.axes import get_regular_master
+
+
+logger = logging.getLogger(__name__)
+
+
+def resolve_intermediate_components(font):
+    components_with_intermediate_layers = set()
+    all_intermediate_locations = set()
+    for glyph in font.glyphs:
+        for layer in glyph.layers:
+            if "coordinates" in layer.attributes:
+                # First, let's find glyphs with intermediate layers
+                # which have components which don't have intermediate layers
+                for shape in layer.components:
+                    ensure_component_has_sparse_layer(font, shape, layer)
+                # Later we will check if everyone who uses me as a component
+                # has the same intermediate layers as I do
+                components_with_intermediate_layers.add(glyph.name)
+                all_intermediate_locations.add(tuple(layer.attributes["coordinates"]))
+
+
+def simple_variation_model(font):
+    tags = [axis.axisTag for axis in font.axes]
+    locations = [x.axes for x in font.masters]
+    limits = {tag: (min(x), max(x)) for tag, x in zip(tags, (zip(*locations)))}
+    master_locations = []
+    default_location = get_regular_master(font).axes
+    for loc in locations:
+        this_loc = {}
+        for ix, axisTag in enumerate(tags):
+            axismin, axismax = limits[axisTag]
+            this_loc[axisTag] = normalizeValue(
+                loc[ix], (axismin, default_location[ix], axismax)
+            )
+        master_locations.append(this_loc)
+    return VariationModel(master_locations, axisOrder=tags), limits
+
+
+def ensure_component_has_sparse_layer(font, component, parent_layer):
+    tags = [axis.axisTag for axis in font.axes]
+    model, limits = simple_variation_model(font)
```

Comment:
the simple model is only taking into account the global master locations.. should it not also consider the locations of the are defined for intermediate layers to be able to interpolate the ones that are missing?

Reply:
IIUC, the problem here is handling a situation whereby there's an intermediate layer that contains a component, and the component's glyph does not have a corresponding intermediate layer explicitly defined at the same location. But this is not necessarily all or nothing, i.e. the component glyph either defining the same set of intermediate layers as the parent layer, or having no intermediate layer at all (only proper master layers). It is possible that the component glyph's layers contain _some_ intermediate while missing others; and when you go to interpolate the latter, should you not also take into account the intermediate layers that _are_ present? I believe so, however right now you only use the master layers from the font.

Reply:
> What we're checking is it if the component defines the same alternate layer as the thing using the component.

yes, that's what I meant with "the component's glyph does not have a corresponding intermediate layer explicitly defined at the same location [as the intermediate layer that contains that component]", slightly more verbosely

> you're right, we should use any intermediate layers that it does have in our interpolation model.

👍